### PR TITLE
bpo-34569: Fix subinterpreter 32-bit  ABI, pystate.c/_new_long_object()

### PR DESCRIPTION
**Pull request by aixtools**
The Lib/test/test__xxsubinterpreters.py test fails in:
ShareableTypeTests.test_int() with the value "-1" when in 32-bit mode (long is 32-bit).

https://bugs.python.org/issue34569

### CODE REVIEW

File: Lib/test/test_embed.py [hunk 1]
```diff
@@ -293,6 +295,49 @@ class InitConfigTests(EmbeddingTestsMixin, unittest.TestCase):
         '_check_hash_pycs_mode': 'default',
         '_frozen': 0,
     }
+    if AIX:
```
Comment (ericsnowcurrently) [inline Lib/test/test_embed.py]: How does it relate to fixing bpo-34569?  Is this a drive-by change?  That would be okay, but otherwise it isn't clear how this relates to the BPO issue.

UPDATE: ah, it looks like you've combined two issues into one PR.  It's better to address different BPO issues in different PRs.

File: Lib/test/test_embed.py [hunk 2]
```diff
@@ -293,6 +295,49 @@ class InitConfigTests(EmbeddingTestsMixin, unittest.TestCase):
...
+        DEFAULT_CONFIG = {
```
Comment (ericsnowcurrently) [inline Lib/test/test_embed.py]: It might be more useful if you update DEFAULT_CONFIG rather than replacing it.  That way we can see how AIX differs from typical for the init config.

File: Lib/test/test__xxsubinterpreters.py [hunk 3]
```diff
@@ -8,12 +8,15 @@
...
+import platform
 
 from test import support
 from test.support import script_helper
-
+from ctypes import c_void_p, sizeof
 
 interpreters = support.import_module('_xxsubinterpreters')
+AIX = platform.system() == 'AIX'
+ABI = sizeof(c_void_p) * 8
```
Comment (ericsnowcurrently) [inline Lib/test/test__xxsubinterpreters.py]: When we're talking about the C-API, "ABI" has a specific meaning, so a better name for this might be `POINTER_SIZE`.

Also, I'd recommend against using ctypes here.  You can get the platform bitness (i.e. size of a pointer) elsewhere in the stdlib.  For example, you could use [`struct.calcsize()`](https://docs.python.org/3/library/struct.html#struct.calcsize), with a [format string of `'P'`](https://docs.python.org/3/library/struct.html#format-characters):

```python
import struct
struct.calcsize('P')  # should return 4 for 32-bit OS, 8 for 64-bit
```

You could use [sys.maxsize](https://docs.python.org/3/library/sys.html#sys.maxsize) too.

Finally, you probably don't need a constant for this (or `AIX`).  Since it's only used in one spot I'd recommend that you move the line there:

```python
    is32bit = sys.maxsize < 2**31
    if platform.system() == 'AIX' and is32bit:
        ...
```

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
I've left some comments below, but there are a couple of high-level things to address.

First, each PR should correspond to one BPO issue.  It looks like this PR has fixes for two issues.

Second, I'd expect the fix to involve changes to `_long_shared()` and/or `_new_long_object()` (in `Python/pystate.c`) or even `PyLong_AsLongLong()`/`PyLong_FromLongLong`/`ULONG_MAX`.  That said, I'm no expert in this platform-specific C-level stuff, so it may be worth getting feedback from someone else too. :)

File: Lib/test/test_embed.py (hunk 1 above)
Comment (aixtools) [reply Lib/test/test_embed.py]: Yes, I made a mistake here with git. I'll remove the DEFAULT_CHANGE stuff. That is a different PR.

Will come later. Thanks for the review!

**Comment by aixtools** [comment]
I have made the requested changes; please review again.

**Comment by ericsnowcurrently** [comment]
Thanks for making those changes.  I'm happy with the code.

That said, as implied above I would expect the test to pass as-is.  `-1` is a legitimate value to pass through a channel.  The fact that the test is failing means that it's doing its job: letting us know when there's a bug. :)  Most likely, either the code in Python/pystate.c (`_long_shared()` and/or `_new_long_object()`) or the implementation of `PyLong_AsLongLong()`/`PyLong_FromLongLong()`/`ULONG_MAX` needs to be fixed.

Consequently, this PR should probably not be merged.  If there's any urgency to getting a passing test suite then I'd recommend making a PR to skip the test under AIX (as a temporary stopgap).  Either way, the BPO issue should stay open until the underlying bug is resolved.

**Comment by aixtools** [comment]
On 19/09/2018 17:53, Eric Snow wrote:
> Thanks for making those changes.  I'm happy with the code.
>
> That said, as implied above I would expect the test to pass as-is.  `-1` is a legitimate value to pass through a channel.  The fact that the test is failing means that it's doing its job: letting us know when there's a bug. :)  Most likely, either the code in Python/pystate.c (`_long_shared()` and/or `_new_long_object()`) or the implementation of `PyLong_AsLongLong()`/`PyLong_FromLongLong()`/`ULONG_MAX` needs to be fixed.
>
> Consequently, this PR should probably not be merged.  If there's any urgency to getting a passing test suite then I'd recommend making a PR to skip the test under AIX (as a temporary stopgap).  Either way, the BPO issue should stay open until the underlying bug is resolved.
>
Understood. Would it "work" to rename to a different issue number (29970
is an issue already opened for 'test' passing.

As to making it work properly, I am not comfortable enough with
internals to actually make a change - but I will make the time later to
find the line where it goes wrong (basically where the value it is
passed is assigned to the "void".

The 'solution' will be something along the line of

#if 32-bit

do something that typecasts from 32-bit to 64-bit. I do not have enough
experience/knowledge to know if type casting from 32-bit "int" to 64-bit
"int" will work in all cases. As to the test, yes, -1 is a valid value.
-2 would fail - rather any negative value - would fail in the same way.

So, as to my question above - if that is okay, should all the other
tests be passing then I'll change this issue number to 29970. But if I
(better we) have made progress with a real solution we can merge those
changes instead (and revert the test back to original).

Michael

**Comment by aixtools** [comment]
afaik - there is a red X because there were issues with the Linux-PR systems. Is there a recognized way to rerun the tests?

**Comment by aixtools** [comment]
@ericsnowcurrently  I have made the requested changes; please review again.

Now a change is pystate.c/new_long_object()

File: Lib/test/test__xxsubinterpreters.py [hunk 4]
```diff
@@ -12,10 +12,8 @@
 from test import support
 from test.support import script_helper
 
-
```
Comment (serhiy-storchaka) [inline Lib/test/test__xxsubinterpreters.py]: Unrelated changes?

File: Misc/NEWS.d/next/Tests/2018-09-09-14-36-59.bpo-34569.okj1Xh.rst [hunk 5]
```diff
@@ -0,0 +1,3 @@
+fix pystate.c/ _new_long_object() so that a 32-bit signed object is returned
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Tests/2018-09-09-14-36-59.bpo-34569.okj1Xh.rst]: Please start sentences with upper letter and add punctuation.

File: Python/pystate.c [hunk 6]
```diff
@@ -1426,7 +1426,12 @@ _str_shared(PyObject *obj, _PyCrossInterpreterData *data)
 static PyObject *
 _new_long_object(_PyCrossInterpreterData *data)
 {
+#if ( ULONG_MAX == UINT_MAX )
+    return PyLong_FromLongLong((int32_t)(data->data));
```
Comment (serhiy-storchaka) [inline Python/pystate.c]: Why not use just `PyLong_FromVoidPtr()`?

File: Lib/test/test__xxsubinterpreters.py (hunk 4 above)
Comment (aixtools) [reply Lib/test/test__xxsubinterpreters.py]: just a line i deleted - by accident. There used to be two blank lines.

File: Python/pystate.c (hunk 6 above)
Comment (aixtools) [reply Python/pystate.c]: a) because the 64-bit version uses PyLong_FromLongLong((int64_t)(data->data)) and that was already working asif it is was a 32-bit "void" or "unsigned" although the test that identified this change is looking at the value -1. Without the change it becomes MAX_INT32U (or something like that, as constant)

File: Misc/NEWS.d/next/Tests/2018-09-09-14-36-59.bpo-34569.okj1Xh.rst (hunk 5 above)
Comment (aixtools) [reply Misc/NEWS.d/next/Tests/2018-09-09-14-36-59.bpo-34569.okj1Xh.rst]: After 12 hours of debugging and testing, sorry I did not notice that I missed he shift key. I'll post an updated PR when I have time.

**Comment by aixtools** [comment]
@ericsnowcurrently - Hi! Have you had time to check this is what you intended for me to do?

File: Python/pystate.c (hunk 6 above)
Comment (kadler) [inline Python/pystate.c]:
```suggestion
    return PyLong_FromLongLong((intptr_t)(data->data));
```

With the above change, there is no need for the `#if` check at all, since `intptr_t` is signed and should be the same size as `void*` on all platforms.

**Comment by kadler** [comment]
@ericsnowcurrently This seems to be a sign extension bug.

`_PyCrossInterpreterData::data` is a `void*` which is holding the integer value 0xFFFFFFFF. As a signed integer it is -1, but as an unsigned integer it is 4294967295. `PyLong_FromLongLong` takes a `long long`, so the compiler will implicitly cast `void*` to `long long`. On 32-bit systems, this will mean that the size of the type will be increased and that means that the signedness of the types will come in to play in determining sign extension.

It seems that on this platform/compiler, `void*` is treated as an `unsigned int`, so there is no sign extension done when casting to a `long long`, eg. 0xFFFFFFFF -> 0x00000000FFFFFFFF. Casting first to a signed integer of the same size as a pointer will fix this. You can validate with this sample program:

```c
#include <stdio.h>

int main()
{
   printf("%lld\n", (long long)(void*)-1);
   return 0;
}
```

With GCC:
```text
-bash-4.3$ gcc -maix32 testi.c && ./a.out
-1
```

With XLC:
```
-bash-4.3$ xlc -q32 testi.c && ./a.out 
4294967295
```

The code as-is will work, but `intptr_t` can be used for this instead of the `#if` condition to keep the code cleaner.

**Comment by kadler** [comment]
There should probably also be a comment added to state why [PyLong_FromVoidPtr](https://docs.python.org/3/c-api/long.html#c.PyLong_FromVoidPtr) was not used, since that would seem most appropriate. The problem is that PyLong_FromVoidPtr explicitly forces the pointer to be unsigned, so you'd never get -1.

File: Python/pystate.c (hunk 6 above)
Comment (aixtools) [reply Python/pystate.c]: nice find. Thx!

**Comment by ncoghlan** [comment]
The actual code change here now looks good to me, but I'm not sure where the NEWS entry should go given that the cross-interpreter data channels aren't officially a public API yet.

@ericsnowcurrently perhaps leaving the entry in Tests makes the most sense right now, even though the code is actually part of the builtins?

**Review by ncoghlan — APPROVED** [review]

**Comment by ericsnowcurrently** [comment]
tl;dr I'll merge this Wednesday Jan 9 (or by the 11th of the latest), unless someone tells me that we could lose info by using `intptr_t`.

Thanks, all, for diving into this.  The solution mostly looks good to me.  Further, my experience with C intrinsic types at the compiler level is small so I'm willing to defer to Nick's judgement here. :)

My only concern is that types match up so there is no loss of information.  In the code in pystate.c, the conversions follow this flow:

pack:
* PyLong_AsLongLong():  PyLongObject -> long long
* cast (implicit):  long long -> int64_t
* cast:  int64_t -> void *

unpack:
* cast:  void * -> int64_t
* cast (implicit):  int64_t -> long long
* PyLong_FromLongLong():  long long -> PyLongObject

The code relies on the equivalency of `int64_t`, `long long`, and `void *`.  It sounds like the equivalency is not perfect on all platforms.  The suggested solution further relies on the equivalency of `intptr_t` and the others.  As long as it lines up I'm okay with it.  The key guarantee is that `sizeof(long long) <= sizeof(void *)` and that any intermediate types are at least as large as `void *`.  If that's true of `intptr_t` (and it should be) then we're good here and we can merge this change. :)

If I don't here any objections by next Wednesday then I'll merge it then.

**Review by ericsnowcurrently — APPROVED** [review]
